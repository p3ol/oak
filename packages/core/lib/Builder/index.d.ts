import {
  AddonObject,
  ElementObject,
  BuilderOptions,
  Component,
  ComponentObject,
  ComponentsGroup,
  ComponentOverride,
  Field,
  FieldObject,
  FieldOverride,
  TextsSheet,
  TextsSheetObject,
  ComponentSettingsField,
  SettingOverride,
  ComponentOverrideObject,
  ElementId,
  ComponentSettingsFieldKeyTuple,
  GetTextCallback,
  ComponentSettingsTab,
} from '../types';
import { Logger } from '../Logger';
import { Emitter, EmitterCallback } from '../Emitter';

export declare class Builder extends Emitter {
  constructor(opts?: {
    addons?: AddonObject[];
    content?: ElementObject[];
    options?: BuilderOptions;
  });

  options: BuilderOptions;
  logger: Logger;

  subscribe(cb: EmitterCallback): (() => void);
  setAddons(addons: AddonObject[]): void;
  addAddon(addon: AddonObject): void;
  removeAddon(addon: AddonObject): void;
  getAvailableComponents(): ComponentsGroup[];
  getComponent(type: string): Component;
  getComponentDisplayableSettings(
    element: ElementObject,
    opts?: {
      component?: Component | ComponentObject;
      fields?: (Field | FieldObject)[];
    },
  ): ComponentSettingsField[];
  getAvailableFields(): Array<Field>;
  getField(type: string): Field;
  getOverride(
    type: string,
    target: string,
    opts?: {
      output?: 'field';
      setting?: ComponentSettingsField;
    },
  ): ComponentOverride | FieldOverride;
  mergeOverrides(
    overrides: (ComponentOverride | FieldOverride | SettingOverride)[]
  ): void;
  getContent(): ElementObject[];
  setContent(content: ElementObject[], opts?: { emit?: boolean }): void;
  createElement(type: string, opts: {
    component?: Component | ComponentObject;
    override?: ComponentOverride | ComponentOverrideObject;
    baseElement?: ElementObject;
    resetIds?: boolean;
  }): object;
  addElement(element: ElementObject, opts?: {
    component?: Component | ComponentObject;
    parent?: ElementObject[];
    position?: 'before' | 'after';
  }): Element;
  addElements(elements: ElementObject[], options?: {
    parent?: ElementObject[];
    position?: 'before' | 'after';
  }): ElementObject[];
  getElement(id: ElementId): ElementObject;
  removeElement(id: ElementId, opts?: {
    deep?: boolean;
  }): boolean;
  setElement(id: ElementId, updates: ElementObject, opts?: {
    deep?: boolean;
  }): ElementObject;
  moveElement(
    element: ElementObject,
    sibling: ElementObject,
    opts?: {
      deep?: boolean;
      position?: 'before' | 'after';
      parent: ElementObject;
    }
  ): ElementObject;
  duplicateElement(
    element: ElementObject,
    opts?: {
      deep?: boolean;
    }
  ): ElementObject;
  getElementSettings(
    element: ElementObject,
    key: string | string[] | ComponentSettingsFieldKeyTuple[],
    def?: any
  ): any;
  canUndo(): boolean;
  canRedo(): boolean;
  undo(): void;
  redo(): void;
  resetHistory(): void;
  generateId(): string;
  getTextSheet(id: string): TextsSheet;
  getText(key: string | GetTextCallback, def?: any): any;
  setText(key: string, value: any): void;
  getActiveTextSheet(): TextsSheet;
  setActiveTextSheet(id: string): void;
  getAvailableSettings(): (ComponentSettingsTab | ComponentSettingsField)[];
}
