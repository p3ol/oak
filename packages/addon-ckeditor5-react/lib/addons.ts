import type { ReactFieldObject, AddonObject } from '@oakjs/react';
import { omit } from '@oakjs/react';
import ClassicEditor from '@oakjs/ckeditor5-build-custom';

import Field, { CKEditorFieldProps } from './Field';

export const ckeditorField = ({
  config,
  editor,
  ...props
}: CKEditorFieldProps = {}): ReactFieldObject => ({
  type: 'ckeditor',
  render: Field,
  ...omit(props, ['onChange']),
  props: {
    editor: editor || ClassicEditor,
    config: {
      ...config,
      link: {
        decorators: {
          openInNewTab: {
            mode: 'manual',
            label: 'Open in a new tab',
            defaultValue: false,
            attributes: {
              target: '_blank',
              rel: 'noopener noreferrer',
            },
          },
        },
      },
      toolbar: {
        items: [
          'undo',
          'redo',
          '|',
          'bold',
          'italic',
          'strikethrough',
          'underline',
          'link',
          'fontColor',
          'fontSize',
          'bulletedList',
          'numberedList',
          '|',
          'removeFormat',
          '|',
          'outdent',
          'indent',
          'alignment',
          '|',
          'horizontalLine',
          'blockQuote',
          'insertTable',
        ],
      },
    },
  },
});

export const ckeditorFieldAddon = (
  props?: CKEditorFieldProps
): AddonObject => ({
  fields: [ckeditorField(props)],
});
