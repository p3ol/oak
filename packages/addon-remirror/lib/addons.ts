import type { AddonObject, ReactFieldObject } from '@oakjs/react';
import {
  BoldExtension,
  FontSizeExtension,
  ItalicExtension,
  UnderlineExtension,
  LinkExtension,
  TextColorExtension,
  NodeFormattingExtension,
} from 'remirror/extensions';

import type { Extensions } from './types';
import Field, { type RemirrorFieldProps } from './Field';

export const remirrorField = ({
  extensions,
  ...props
}: RemirrorFieldProps = {}): ReactFieldObject => ({
  type: 'remirror',
  render: Field,
  ...props,
  props: {
    extensions,
  },
});

export const basicExtensions = (): Extensions[] => [
  new BoldExtension({}),
  new ItalicExtension(),
  new UnderlineExtension(),
  new LinkExtension({}),
  new TextColorExtension({}),
  new FontSizeExtension({
    defaultSize: '16px',
    unit: 'px',
  }),
  new NodeFormattingExtension({}),
];

export const remirrorFieldAddon = (props?: ReactFieldObject): AddonObject => ({
  fields: [remirrorField({
    extensions: basicExtensions,
    ...props,
  })],
});
