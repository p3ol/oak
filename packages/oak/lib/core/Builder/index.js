import { useRef } from 'react';
import { classNames } from '@poool/junipero-utils';

import { useOptions, useBuilder } from '../../hooks';
import Element from '../Element';
import Catalogue from '../Catalogue';
import Icon from '../Icon';

export default () => {
  const catalogueRef = useRef();
  const {
    components,
    content,
    addElement,
    _setSettingsHolderRef,
    undo,
    redo,
    isUndoPossible,
    isRedoPossible,
    getText,
    getOverrides,
  } = useBuilder();
  const { debug, historyButtonsEnabled } = useOptions();

  const onAppend = component => {
    const element = component.construct?.() || {};
    const overrides = getOverrides('component', element.type);
    const elementWithContent = {
      ...element,
      content: typeof element.content === 'function'
        ? element.content(getText) : element.content,
    };
    addElement({
      ...elementWithContent,
      ...(
        overrides?.construct &&
        typeof overrides.construct === 'function'
          ? overrides.construct(elementWithContent)
          : {}
      ),
    });
    catalogueRef.current?.close();
  };

  const onPaste = element => {
    addElement(element, { normalizeOptions: { resetIds: true } });
    catalogueRef.current?.close();
  };

  return (
    <div className="oak-builder">
      { historyButtonsEnabled !== false && (
        <div className="oak-undo-redo">
          <a onClick={() => isUndoPossible && undo()}
            className={classNames('oak-undo', !isUndoPossible && 'disabled')}
          >
            <Icon>undo</Icon>
          </a>
          <a onClick={() => isRedoPossible && redo()}
            className={classNames('oak-redo', !isRedoPossible && 'disabled')}
          >
            <Icon>redo</Icon>
          </a>
        </div>
      ) }

      { content.map((item, i) => (
        <Element
          key={item.id || i}
          index={i}
          element={item}
        />
      )) }

      <div className="oak-add-element">
        <Catalogue
          ref={catalogueRef}
          onAppend={onAppend}
          onPaste={onPaste}
        />
      </div>

      <div ref={_setSettingsHolderRef} />

      { debug && (
        <pre>
          <p>Content:</p>
          { JSON.stringify(content, null, 2) }

          <p>Components:</p>
          { JSON.stringify(components, null, 2) }
        </pre>
      )}
    </div>
  );
};
