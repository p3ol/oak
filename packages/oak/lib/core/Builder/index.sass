@use "colors" as colors

.oak-builder
  .oak-add-element
    text-align: center

  .oak-undo-redo
    width:100%
    display: flex
    padding-bottom: 10px

    .oak-undo, .oak-redo
      color: var(--oak-text-color)
      cursor: pointer
      align-items: center
      padding: 10px 15px 6px
      background: colors.$black-squeeze
      border-radius: 3px
      user-select: none

      &.disabled
        opacity: 0.5
        cursor: default

      &:hover
        &:not(.disabled)
          box-shadow: 0 0 0 1px var(--oak-main-color)

    .oak-undo
      margin-right: 10px
