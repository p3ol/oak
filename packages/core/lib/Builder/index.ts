import { exists } from '@junipero/core';
import { v4 as uuid } from 'uuid';

import type {
  AddonObject,
  BuilderObject,
  ComponentObject,
  ComponentOverrideObject,
  ComponentSettingsFieldObject,
  ComponentSettingsTabObject,
  ComponentsGroupObject,
  ElementId,
  ElementObject,
  ElementSettingsKeyObject,
  EventCallback,
  FieldOverrideObject,
  GetTextCallback,
} from '../types';
import {
  BuilderOptions,
  Component,
  ComponentOverride,
  FieldOverride,
} from '../classes';
import Emitter from '../Emitter';
import Logger from '../Logger';
import Components from '../Components';
import Fields from '../Fields';
import Overrides from '../Overrides';
import Store from '../Store';
import Texts from '../Texts';
import Settings from '../Settings';

export default class Builder extends Emitter {
  logger: Logger = null;
  options: BuilderOptions = null;

  #components: Components = null;
  #fields: Fields = null;
  #overrides: Overrides = null;
  #texts: Texts = null;
  #store: Store = null;
  #settings: Settings = null;
  #addons: AddonObject[] = [];

  constructor (opts: {
    addons?: AddonObject[],
    content?: ElementObject[],
    options?: BuilderObject
  } = {}) {
    super();

    this.options = new BuilderOptions(opts.options);
    this.logger = new Logger({ builder: this });

    this.#components = new Components({ builder: this });
    this.#fields = new Fields({ builder: this });
    this.#overrides = new Overrides({ builder: this });
    this.#store = new Store({ builder: this });
    this.#texts = new Texts({ builder: this });
    this.#settings = new Settings({ builder: this });

    if (Array.isArray(opts.addons)) {
      this.#addons = opts.addons;
      opts.addons.forEach(addon => {
        this.logger.log('Initializing builder with addon:', addon);
        this.addAddon(addon);
      });
    }

    if (opts.content) {
      this.logger.log('Initializing builder with content:', opts.content);
      this.#store.set(opts.content);
    }
  }

  subscribe (callback: EventCallback) {
    const subscriptions = [
      super.subscribe(callback),
      this.#store.subscribe(this.emit.bind(this)),
      this.#texts.subscribe(this.emit.bind(this)),
      this.#components.subscribe(this.emit.bind(this)),
      this.#fields.subscribe(this.emit.bind(this)),
      this.#overrides.subscribe(this.emit.bind(this)),
      this.#settings.subscribe(this.emit.bind(this)),
    ];

    return () => {
      subscriptions.forEach(u => u());
    };
  }

  setAddons (addons: AddonObject[]): void {
    this.#addons?.forEach(addon => {
      this.logger.log('Removing builder addon:', addon);
      this.removeAddon(addon);
    });

    this.#addons = addons;
    this.#addons?.forEach(addon => {
      this.logger.log('Updating builder addon:', addon);
      this.addAddon(addon);
    });

    this.emit('addons.update', addons);
  }

  addAddon (addon: AddonObject): void {
    addon.fields?.forEach(field => {
      this.#fields.add(field);
    });

    addon.components?.forEach(component => {
      this.#components.add(component);
    });

    addon.texts?.forEach(sheet => {
      this.#texts.addSheet(sheet);
    });

    addon.overrides?.forEach(override => {
      this.#overrides.add(override);
    });

    addon.settings?.forEach(setting => {
      this.#settings.add(setting);
    });
  }

  removeAddon (addon: AddonObject): void {
    addon.settings?.forEach(setting => {
      this.#settings.remove(setting.id);
    });

    addon.overrides?.forEach(override => {
      this.#overrides.remove(override.id);
    });

    addon.texts?.forEach(sheet => {
      this.#texts.removeSheet(sheet.id);
    });

    addon.components?.forEach(component => {
      this.#components.remove(component.id);
    });

    addon.fields?.forEach(field => {
      this.#fields.remove(field.type);
    });
  }

  getAvailableComponents (): ComponentsGroupObject[] {
    const {
      groups,
      defaultGroup,
    } = this.#components.getAll();

    return [...groups, defaultGroup];
  }

  getComponent (type: string): ComponentObject {
    return this.#components.getComponent(type);
  }

  getComponentDisplayableSettings (
    element: ElementObject,
    { component, override }: {
      component: ComponentObject;
      override?: ComponentOverrideObject | ComponentOverride;
    },
  ): (ComponentSettingsTabObject | ComponentSettingsFieldObject)[] {
    const component_ = new Component(component);
    const override_ = new ComponentOverride(override ?? {});

    return [
      ...this.#components
        .getDisplayableSettings?.(
          element, { component: component_, override: override_ }
        ) || [],
      ...this.#settings.getDisplayable?.(element) || [],
    ];
  }

  getAvailableFields () {
    return this.#fields.all();
  }

  getField (type: string) {
    return this.#fields.get(type);
  }

  getOverride (
    type: 'component' | 'field' | 'setting',
    target: string,
    opts?: any
  ) {
    return this.#overrides.get(type, target, opts);
  }

  mergeOverrides (
    overrides: (ComponentOverrideObject | FieldOverrideObject)[]
  ) {
    const ovrrides_ = overrides.map(override => {
      if (override.type === 'component') {
        return new ComponentOverride(override);
      } else {
        return new FieldOverride(override);
      }
    });

    return this.#overrides.merge(ovrrides_);
  }

  getContent () {
    return this.#store.get();
  }

  setContent (content: ElementObject[], options: { emit?: boolean }) {
    this.#store.set(content, options);
  }

  createElement (type: string, options?: {
    component?: ComponentObject | Component,
    override?: ComponentOverrideObject,
    baseElement?: ElementObject,
    [_: string]: any
  }) {
    return this.#store.createElement(type, options);
  }

  addElement (element: ElementObject, options?: {
    parent?: ElementObject[],
    position?: 'before' | 'after',
    component?: ComponentObject,
    [_: string]: any
  }) {
    return this.#store.addElement(element, options);
  }

  addElements (elements: ElementObject[], options?:{
    parent?: ElementObject[],
    position?: 'before' | 'after',
    component?: ComponentObject,
    [_: string]: any
  }) {
    return this.#store.addElements(elements, options);
  }

  getElement (id: ElementId, options?: {
    parent?: ElementObject[];
    deep?: boolean;
  }) {
    return this.#store.getElement(id, options);
  }

  removeElement (id: ElementId, options?: {
    parent?: ElementObject[];
    deep?: boolean;
  }) {
    return this.#store.removeElement(id, options);
  }

  setElement (id: ElementId, updates: ElementObject, options?: {
    element?: ElementObject;
    parent?: ElementObject[];
    deep?: boolean;
  }) {
    return this.#store.setElement(id, updates, options);
  }

  moveElement (
    element: ElementObject,
    sibling: ElementObject,
    options?: {
      parent?: ElementObject[],
      position?: 'before' | 'after'
    }
  ) {
    return this.#store.moveElement(element, sibling, options);
  }

  duplicateElement (element: ElementObject, options?: {
    parent?: ElementObject[];
  }) {
    return this.#store.duplicateElement(element, options);
  }

  getElementSettings (
    element: ElementObject,
    key: ElementSettingsKeyObject | string,
    def?: any
  ) {
    return this.#store.getElementSettings(element, key, def);
  }

  setElementSettings (
    element: ElementObject,
    key: ElementSettingsKeyObject | string,
    value: any
  ) {
    return this.#store.setElementSettings(element, key, value);
  }

  undo () {
    this.#store.undo();
  }

  redo () {
    this.#store.redo();
  }

  canUndo () {
    return this.#store.canUndo();
  }

  canRedo () {
    return this.#store.canRedo();
  }

  resetHistory () {
    this.#store.resetHistory();
  }

  generateId () {
    const customId = this.options.generateId?.();

    return exists(customId) && customId !== '' ? customId : uuid();
  }

  getTextSheet (id: string) {
    return this.#texts.getSheet(id);
  }

  getText (key: GetTextCallback | string, def?: string) {
    return this.#texts.get(key, def);
  }

  setText (key: string, value: string) {
    return this.#texts.set(key, value);
  }

  getActiveTextSheet () {
    return this.#texts.getActiveSheet();
  }

  setActiveTextSheet (id: string) {
    return this.#texts.setActiveSheet(id);
  }

  getAvailableSettings () {
    return this.#settings.all();
  }

  setOptions (opts: Partial<BuilderObject>) {
    this.options = new BuilderOptions(opts);
    this.emit('options.update', this.options);
  }
}
