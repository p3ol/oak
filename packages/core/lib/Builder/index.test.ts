import { vi } from 'vitest';

import type { ElementObject } from '../types';
import { baseAddon } from '../addons';
import Builder from './index';

describe('Builder', () => {
  it('should create a builder', () => {
    const content: ElementObject[] = [
      { type: 'title', content: 'This is a title' },
      { type: 'text', content: 'This is a text' },
    ];

    let id = 0;
    const builder = new Builder({
      addons: [baseAddon()],
      content,
      options: {
        generateId: () => id++,
      },
    });

    expect(builder.getContent()).toMatchSnapshot('Base content');

    builder.removeElement(1);

    expect(builder.getContent()).toMatchSnapshot('Without text');

    const button = { type: 'button', content: 'This is a button' };
    builder.addElement(button);

    expect(builder.getContent()).toMatchSnapshot('With button');

    builder.setElement(2, { content: 'This is a new button' });

    expect(builder.getContent()).toMatchSnapshot('With button updated');

    builder.moveElement(button, content[0], { position: 'before' });

    expect(builder.getContent()).toMatchSnapshot('With button & title swapped');

    builder.duplicateElement(button);

    expect(builder.getContent()).toMatchSnapshot('With button duplicated');
  });

  it('should allow to listen to changes', () => {
    const builder = new Builder({
      addons: [baseAddon()],
    });

    const listener = vi.fn();
    const unsubscribe = builder.subscribe(listener);

    builder.addElement({ type: 'title', content: 'This is a title' });

    expect(listener)
      .toHaveBeenCalledWith('content.update', expect.arrayContaining([
        expect.objectContaining({ type: 'title', content: 'This is a title' }),
      ]));

    unsubscribe();

    builder.addElement({ type: 'text', content: 'This is a text' });

    // 1 -> content.update
    // 2 -> history.commit
    expect(listener).toHaveBeenCalledTimes(2);
  });

  it('should allow to add addons', () => {
    const builder = new Builder();

    builder.addAddon({
      components: [{
        id: 'richtext',
      }],
      fields: [{
        type: 'richtext-field',
      }],
    });

    expect(builder.getAvailableComponents()).toMatchSnapshot('Components');
    expect(builder.getAvailableFields()).toMatchSnapshot('Fields');
  });

  it('should allow to translate texts', () => {
    const builder = new Builder({
      addons: [{
        texts: [{
          id: 'fr',
          texts: {
            core: {
              pasteFromClipboard: 'Coller depuis le presse-papier',
            },
          },
        }],
      }],
    });

    expect(builder.getText('core.pasteFromClipboard', 'Paste from clipboard'))
      .toBe('Paste from clipboard');

    builder.setActiveTextSheet('fr');

    expect(builder.getText('core.pasteFromClipboard', 'Paste from clipboard'))
      .toBe('Coller depuis le presse-papier');
  });
});
