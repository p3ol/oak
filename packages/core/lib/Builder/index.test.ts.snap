// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Builder > should allow to add addons > Components 1`] = `
[
  ComponentsGroup {
    "components": [
      Component {
        "construct": undefined,
        "copyable": true,
        "deserialize": undefined,
        "disallow": [],
        "draggable": true,
        "droppable": true,
        "duplicable": true,
        "duplicate": undefined,
        "editable": true,
        "getContainers": undefined,
        "group": undefined,
        "hasCustomInnerContent": false,
        "icon": undefined,
        "id": "richtext",
        "name": "",
        "options": [],
        "render": undefined,
        "sanitize": undefined,
        "serialize": undefined,
        "settings": ComponentSettingsForm {
          "defaults": {},
          "fields": [],
          "floatingSettings": undefined,
          "title": undefined,
        },
        "type": "component",
        "usable": true,
      },
    ],
    "id": "other",
    "name": [Function],
    "type": "group",
    "usable": true,
  },
]
`;

exports[`Builder > should allow to add addons > Fields 1`] = `
[
  Field {
    "props": {},
    "render": undefined,
    "type": "richtext-field",
  },
]
`;

exports[`Builder > should create a builder > Base content 1`] = `
[
  {
    "content": "This is a title",
    "headingLevel": "h1",
    "id": 0,
    "type": "title",
  },
  {
    "content": "This is a text",
    "id": 1,
    "type": "text",
  },
]
`;

exports[`Builder > should create a builder > With button & title swapped 1`] = `
[
  {
    "content": "This is a new button",
    "id": 2,
    "type": "button",
  },
  {
    "content": "This is a title",
    "headingLevel": "h1",
    "id": 0,
    "type": "title",
  },
]
`;

exports[`Builder > should create a builder > With button 1`] = `
[
  {
    "content": "This is a title",
    "headingLevel": "h1",
    "id": 0,
    "type": "title",
  },
  {
    "content": "This is a button",
    "id": 2,
    "type": "button",
  },
]
`;

exports[`Builder > should create a builder > With button duplicated 1`] = `
[
  {
    "content": "This is a new button",
    "id": 2,
    "type": "button",
  },
  {
    "content": "This is a new button",
    "id": 3,
    "type": "button",
  },
  {
    "content": "This is a title",
    "headingLevel": "h1",
    "id": 0,
    "type": "title",
  },
]
`;

exports[`Builder > should create a builder > With button updated 1`] = `
[
  {
    "content": "This is a title",
    "headingLevel": "h1",
    "id": 0,
    "type": "title",
  },
  {
    "content": "This is a new button",
    "id": 2,
    "type": "button",
  },
]
`;

exports[`Builder > should create a builder > Without text 1`] = `
[
  {
    "content": "This is a title",
    "headingLevel": "h1",
    "id": 0,
    "type": "title",
  },
]
`;
